---
title: "BfimputePaper"
author: "Zi-Hang Wen"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
---

# Before You Begin
## Prepare the packages
`Bfimpute` is a Bayesian matrix factorization imputation method. You can
download it by running:
```{r, eval = FALSE}
devtools::install_github("maiziezhoulab/Bfimpute")
```
If imputation is all you need, you can skip the following installation steps.

`BfimputePaper` is a software integrating scRNA-seq analysis used for paper
[Bfimpute: A Bayesian factorization method to recover single-cell RNA sequencing data](https://doi.org/10.1101/2021.02.10.430649). Some packages we used are only
published in github and some attached functions used in this package may have
different names or parameters from their other versions. In order to
avoid errors, please install the following packages before installing
`BfimputePaper` and check the `sessioninfo()` if some functions do not workout:

```{r, eval = FALSE}
# Splatter
BiocManager::install("splatter")
# scater
BiocManager::install("scater")
# Spectrum
install.packages("Spectrum")
# DESeq2
BiocManager::install("DESeq2")
# Monocle
devtools::install_github("cole-trapnell-lab/monocle-release")
# TSCAN
devtools::install_github("zji90/TSCAN")
# WGCNA
install.packages("WGCNA")
# org.Hs.eg.db
BiocManager::install("org.Hs.eg.db")
# clusterProfiler
BiocManager::install("clusterProfiler")
# TSCAN
devtools::install_github("zji90/TSCAN")
# scDatasets
devtools::install_github("gongx030/scDatasets")
```

Then install `BfimputePaper` with:
```{r, eval = FALSE}
devtools::install_github("maiziezhoulab/BfimputePaper")
```
## Prepare the data
In this protocol, we will use the data from Gene Expression Omnibus (GEO) 
Database with GEO Series accession number GSE75748. So, download the 
'GSE75748_sc_time_course_ec.csv.gz' and 'GSE75748_bulk_time_course_ec.csv.gz'
from [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75748) and
unzip them.

If you want to use your own scRNA-seq data, you can conduct alignment and
generate the count matrix with genes as the rows and cells as the columns.

# Main Protocol
