---
title: "BfimputeProtocol"
author: "Zi-Hang Wen"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    keep_md: true
---

# Before You Begin
## Prepare the expression matrix
If you already have the the scRNA-seq expression matrix with genes as the rows
and cells as the columns, you can skip this section.

We hereby give an example on how to generate the expression matrix from a human
embryonic stem cell study (GEO: GSE75748). In this protocol, we will use the
time course data from 00h to 96h on both single cells and bulk samples. The
SRA accession numbers lists are stored in folder `SRRlist`. To use this protocol on your
own experiment, you can generate your files similarly. For convenience, we also
put the R script `creat_matrix.R` for cell counts merging and quality control
which will be used in future steps.

Firstly, create an empty folder `raw_data` as the workspace and put folder
`SRRlist` in that workspace.
```{bash, eval = FALSE}
mkdir raw_data
mv SRRlist raw_data
cd raw_data
```

### Alignment
Download HISAT2 and its index from [here](http://daehwankimlab.github.io/hisat2/download/).
Then perform alignment for each stage one by one or simultaneously using HISAT2.
```{bash, eval = FALSE}
name="sc00h"    # sc12h sc24h sc36h sc72h sc96h bulk12h bulk24h bulk36h bulk72h bulk96h
file=SRRlist/"$name".txt
sampath=sam_"$name"/

mkdir $sampath

while read line; do
    echo $line
    YOUR_HISAT2_PATH/hisat2 -t \
        -x YOUR_INDEX_PATH/genome \
        --sra-acc $line \
        -S "$sampath"/"$line"_GRCh37.sam
done < $file
```

### Sort
Dowload and install SAMtools from [here](http://www.htslib.org/download/). Then
transform the sam files to bam files, sort the alignments, and create index
using SAMtools.
```{bash, eval = FALSE}
name="sc00h"    # sc12h sc24h sc36h sc72h sc96h bulk12h bulk24h bulk36h bulk72h bulk96h

file=SRRlist/"$name".txt
sampath=sam_"$name"/
bampath=bam_"$name"/

mkdir $bampath
mkdir temp

while read line; do
    echo $line
    samtools view -bS "$sampath"/"$line"_GRCh37.sam | \
        samtools sort -T temp/$line \
        -o "$bampath"/"$line"_GRCh37.bam
    samtools index "$bampath"/"$line"_GRCh37.bam
done < $file
```

### Count
Download and install HTSeq from [here](https://htseq.readthedocs.io/en/master/install.html).
Download the GTF file from [here](http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz)
```{bash, eval = FALSE}
name="sc00h"    # sc12h sc24h sc36h sc72h sc96h bulk12h bulk24h bulk36h bulk72h bulk96h

file=SRRlist/"$name".txt
bampath=bam_"$name"/
countpath=count_"$name"/

mkdir $countpath

while read line; do
    echo $line
    htseq-count -s no -r pos -f bam "$bampath"/"$line"_GRCh37.bam YOUR_GTF_PATH/Homo_sapiens.GRCh37.75.gtf > "$countpath"/"$line"_GRCh37.count
done < $file
```

The next step is merge all the individual single cells to a scRNA-seq count
matrix and bulk samples to a bulk RNA-seq count matrix. Then perform quality
control and align the genes for scRNA-seq count matrix and bulk RNA-seq count
matrix.
```{bash, eval = FALSE}
cd SRRlist
Rscript create_matrix.R
```

## Prepare the packages
`Bfimpute` is a Bayesian matrix factorization imputation method. You can
download it by running:
```{r, eval = FALSE}
devtools::install_github("maiziezhoulab/Bfimpute")
```
If imputation is all you need, you can skip the following installation steps.

`BfimputeProtocol` is a software integrating scRNA-seq analysis used for paper
[Bfimpute: A Bayesian factorization method to recover single-cell RNA sequencing data](https://doi.org/10.1101/2021.02.10.430649). Some packages we used are only
published in github and some attached functions used in this package may have
different names or parameters from their other versions. In order to
avoid errors, please install the following packages before installing
`BfimputeProtocol` and check the `sessioninfo()` if some functions do not workout:

```{r, eval = FALSE}
# Splatter
BiocManager::install("splatter")
# scater
BiocManager::install("scater")
# Spectrum
install.packages("Spectrum")
# DESeq2
BiocManager::install("DESeq2")
# Monocle
devtools::install_github("cole-trapnell-lab/monocle-release")
# TSCAN
devtools::install_github("zji90/TSCAN")
# WGCNA
install.packages("WGCNA")
# org.Hs.eg.db
BiocManager::install("org.Hs.eg.db")
# clusterProfiler
BiocManager::install("clusterProfiler")
# TSCAN
devtools::install_github("zji90/TSCAN")
# scDatasets
devtools::install_github("gongx030/scDatasets")
```

Then install `BfimputeProtocol` with:
```{r, eval = FALSE}
devtools::install_github("maiziezhoulab/BfimputeProtocol")
```

# Main Protocol
